core/src/main/java/org/elasticsearch/indices/cache/query/terms/TermsLookup.java

```diff
@@ -64,13 +72,106 @@ public String getPath() {
         return path;
     }
 
-    @Nullable
-    public QueryParseContext getQueryParseContext() {
-        return queryParseContext;
+    public void index(String index) {
+        this.index = index;
+    }
+
+    public void type(String type) {
+        this.type = type;
+    }
+
+    public void id(String id) {
+        this.id = id;
+    }
+
+    public void routing(String routing) {
+        this.routing = routing;
+    }
+
+    public void path(String path) {
+        this.path = path;
     }
 
     @Override
     public String toString() {
         return index + "/" + type + "/" + id + "/" + path;
     }
+
+    @Override
+    public TermsLookup readFrom(StreamInput in) throws IOException {
+        TermsLookup termsLookup = new TermsLookup();
+        termsLookup.index = in.readOptionalString();
+        termsLookup.type = in.readString();
+        termsLookup.id = in.readString();
+        termsLookup.path = in.readString();
+        termsLookup.routing = in.readOptionalString();
+        return termsLookup;
+    }
+
+    public static TermsLookup readTermsLookupFrom(StreamInput in) throws IOException {
+        return PROTOTYPE.readFrom(in);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeOptionalString(index);
+        out.writeString(type);
+        out.writeString(id);
+        out.writeString(path);
+        out.writeOptionalString(routing);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        if (index != null) {
+            builder.field("index", index);
+        }
+        builder.field("type", type);
+        builder.field("id", id);
+        builder.field("path", path);
+        if (routing != null) {
+            builder.field("routing", routing);
+        }
+        return builder;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(getClass(), index, type, id, routing, path);
```

Comment:
I think we can remove getClass() from here and make this class final to prevent any problem around that